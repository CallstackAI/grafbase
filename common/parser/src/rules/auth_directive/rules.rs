use std::collections::HashSet;

use serde::{Deserialize, Serialize};

use dynaql::ServerError;
use dynaql_value::ConstValue;

use super::operations::Operations;
use crate::VisitorContext;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(tag = "allow")]
#[serde(deny_unknown_fields)]
#[non_exhaustive]
pub enum AuthRule {
    /// Public data access via API keys
    // Ex: { allow: anonymous }
    #[serde(alias = "public")]
    #[serde(rename_all = "camelCase")]
    Anonymous {
        // Note: we don't support operations as our playground needs full access
    },

    /// Signed-in user data access via OIDC
    // Ex: { allow: private }
    //     { allow: private, operations: [create, read] }
    #[serde(rename_all = "camelCase")]
    Private {
        #[serde(default)]
        operations: Operations,
    },

    /// User group-based data access via OIDC
    // Ex: { allow: groups, groups: ["admin"] }
    //     { allow: groups, groups: ["admin"], operations: [update, delete] }
    #[serde(rename_all = "camelCase")]
    Groups {
        #[serde(with = "::serde_with::rust::sets_duplicate_value_is_error")]
        groups: HashSet<String>,

        #[serde(default)]
        operations: Operations,
    },

    /// Owner-based data access via OIDC
    // Ex: { allow: owner }
    //     { allow: owner, operations: [create, read] }
    #[serde(rename_all = "camelCase")]
    Owner {
        #[serde(default)]
        operations: Operations,
    },
}

impl AuthRule {
    pub fn from_value(_ctx: &VisitorContext<'_>, value: &ConstValue) -> Result<Self, ServerError> {
        // We convert the value to JSON to leverage serde for deserialization
        let value = match value {
            ConstValue::Object(_) => value
                .clone()
                .into_json()
                .map_err(|err| ServerError::new(err.to_string(), None))?,
            _ => return Err(ServerError::new("auth rule must be an object", None)),
        };

        let rule: AuthRule =
            serde_json::from_value(value).map_err(|err| ServerError::new(format!("auth rule: {err}"), None))?;

        Ok(rule)
    }
}
