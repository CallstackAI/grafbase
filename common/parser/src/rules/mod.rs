pub mod basic_type;
pub mod check_type_validity;
pub mod check_types_underscore;
pub mod model_directive;
pub mod visitor;
