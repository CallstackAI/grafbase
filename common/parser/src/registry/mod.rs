//! TODO:
//!
//! -> Split each of the creation and add tests with SDL
//!
use case::CaseExt;

use dynaql::indexmap::IndexMap;
use dynaql::registry::enums::DynaqlEnum;
use dynaql::registry::relations::MetaRelation;
use dynaql::registry::transformers::Transformer;
use dynaql::registry::{
    resolvers::dynamo_mutation::DynamoMutationResolver, resolvers::Resolver, resolvers::ResolverType,
    variables::VariableResolveDefinition, MetaField, MetaInputValue, MetaType,
};
use dynaql::registry::{MetaEnumValue, Registry};
use dynaql::validation::dynamic_validators::DynValidator;
use dynaql::AuthConfig;
use dynaql_parser::types::{BaseType, FieldDefinition, ObjectType, TypeDefinition};
use grafbase::auth::Operations;
use std::fmt::Display;

use crate::registry::names::MetaNames;
use crate::rules::length_directive::{LENGTH_DIRECTIVE, MAX_ARGUMENT, MIN_ARGUMENT};
use crate::rules::visitor::VisitorContext;
use crate::utils::{to_input_type, to_lower_camelcase};

mod mutations;
pub mod names;
mod pagination;
mod relations;
mod search;
pub use mutations::{add_mutation_create, add_mutation_update, NumericFieldKind};
pub use pagination::{add_query_paginated_collection, generate_pagination_args};
pub use search::add_query_search;

fn register_dynaql_enum<T: DynaqlEnum>(registry: &mut Registry) -> BaseType {
    let type_name = T::ty().to_string();
    registry.create_type(
        |_| MetaType::Enum {
            name: type_name.clone(),
            description: None,
            enum_values: IndexMap::from_iter(
                T::values()
                    .into_iter()
                    .map(|value| {
                        (
                            value.clone(),
                            MetaEnumValue {
                                name: value,
                                description: None,
                                deprecation: dynaql::registry::Deprecation::NoDeprecated,
                                visible: None,
                                value: None,
                            },
                        )
                    })
                    .collect::<Vec<_>>(),
            ),
            visible: None,
            rust_typename: type_name.clone(),
        },
        &type_name,
        &type_name,
    );
    BaseType::named(&type_name)
}

/// Create an input type for a non_primitive Type.
pub fn add_input_type_non_primitive(ctx: &mut VisitorContext<'_>, object: &ObjectType, type_name: &str) -> String {
    let type_name = type_name.to_string();
    let input_type = format!("{}Input", type_name.to_camel());

    // Input
    ctx.registry.get_mut().create_type(
        |_| MetaType::InputObject {
            name: input_type.clone(),
            description: Some(format!("{type_name} input type.")),
            oneof: false,
            input_fields: {
                let mut input_fields = IndexMap::new();
                for field in &object.fields {
                    let name = &field.node.name.node;

                    input_fields.insert(
                        name.clone().to_string(),
                        MetaInputValue {
                            description: field.node.description.clone().map(|x| x.node),
                            ..MetaInputValue::new(
                                name.to_string(),
                                to_input_type(&ctx.types, field.node.ty.clone().node),
                            )
                        },
                    );
                }
                input_fields
            },
            visible: None,
            rust_typename: input_type.clone(),
        },
        &input_type,
        &input_type,
    );

    input_type
}

/// Add the remove mutation for a given Object
pub fn add_remove_mutation(ctx: &mut VisitorContext<'_>, type_name: &str, auth: Option<&AuthConfig>) {
    let type_name = type_name.to_string();
    let delete_payload_name = format!("{}DeletePayload", type_name.to_camel());

    // DeletePayload
    ctx.registry.get_mut().create_type(
        |_| MetaType::Object {
            name: delete_payload_name.clone(),
            description: None,
            fields: {
                let mut fields = IndexMap::new();
                let name = "deletedId".to_string();
                fields.insert(
                    name.clone(),
                    MetaField {
                        name,
                        description: None,
                        args: Default::default(),
                        // TODO: Should be infered from the entity depending on the directives
                        ty: "ID!".to_string(),
                        deprecation: Default::default(),
                        cache_control: Default::default(),
                        external: false,
                        requires: None,
                        provides: None,
                        visible: None,
                        compute_complexity: None,
                        edges: Vec::new(),
                        relation: None,
                        resolve: None,
                        transformer: Some(Transformer::JSONSelect {
                            property: "id".to_string(),
                        }),
                        plan: None,
                        required_operation: Some(Operations::DELETE),
                        auth: auth.cloned(),
                    },
                );
                fields
            },
            cache_control: Default::default(),
            extends: false,
            keys: None,
            is_node: false,
            visible: None,
            is_subscription: false,
            rust_typename: delete_payload_name.clone(),
            constraints: vec![],
        },
        &delete_payload_name,
        &delete_payload_name,
    );

    // deleteMutation
    ctx.mutations.push(MetaField {
        name: format!("{}Delete", to_lower_camelcase(&type_name)),
        description: Some(format!("Delete a {type_name} by ID or unique field")),
        args: {
            let mut args = IndexMap::new();
            args.insert(
                "by".to_owned(),
                MetaInputValue::new("by", format!("{type_name}ByInput!")),
            );
            args
        },
        ty: delete_payload_name,
        deprecation: dynaql::registry::Deprecation::NoDeprecated,
        cache_control: Default::default(),
        external: false,
        provides: None,
        requires: None,
        visible: None,
        edges: Vec::new(),
        relation: None,
        compute_complexity: None,
        resolve: Some(Resolver {
            id: Some(format!("{}_delete_resolver", type_name.to_lowercase())),
            r#type: ResolverType::DynamoMutationResolver(DynamoMutationResolver::DeleteNode {
                ty: type_name,
                by: VariableResolveDefinition::InputTypeName("by".to_owned()),
            }),
        }),
        plan: None,
        transformer: None,
        required_operation: Some(Operations::DELETE),
        auth: auth.cloned(),
    });
}

fn get_length_validator(field: &FieldDefinition) -> Option<DynValidator> {
    use tuple::Map;
    field
        .directives
        .iter()
        .find(|directive| directive.node.name.node == LENGTH_DIRECTIVE)
        .map(|directive| {
            let (min_value, max_value) = (MIN_ARGUMENT, MAX_ARGUMENT).map(|argument_name| {
                directive.node.get_argument(argument_name).and_then(|argument| {
                    if let dynaql_value::ConstValue::Number(ref min) = argument.node {
                        min.as_u64().and_then(|min| min.try_into().ok())
                    } else {
                        None
                    }
                })
            });
            DynValidator::length(min_value, max_value)
        })
}

/// Used to keep track of the parent relation when created nested input types
/// TODO: Merge it with MetaRelation?
pub struct ParentRelation<'a> {
    /// TypeDefinition of @model type
    model_type_definition: &'a TypeDefinition,
    meta: &'a MetaRelation,
}

impl<'a> Display for ParentRelation<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} relation of {}",
            self.meta.name,
            MetaNames::model(self.model_type_definition)
        )
    }
}
