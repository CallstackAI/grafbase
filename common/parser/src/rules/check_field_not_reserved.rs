//! ### What it does
//!
//! Check that fields are not reserved for Modelized types
//!
//! ### Why?
//!
//! We do not want Modelized type with created_at and updated_at fields as we add them
//! ourselves.
use super::model_directive::MODEL_DIRECTIVE;
use super::visitor::{Visitor, VisitorContext};
use if_chain::if_chain;

pub struct CheckModelizedFieldReserved;

const RESERVED_FIELDS: [&str; 4] = ["created_at", "updated_at", "createdAt", "updatedAt"];

impl<'a> Visitor<'a> for CheckModelizedFieldReserved {
    fn directives(&self) -> String {
        String::new()
    }

    fn enter_field(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        field: &'a dynaql::Positioned<dynaql_parser::types::FieldDefinition>,
        parent: &'a dynaql::Positioned<dynaql_parser::types::TypeDefinition>,
    ) {
        if_chain! {
            let name = &field.node.name.node;
            if RESERVED_FIELDS.contains(&name.as_str());
            if parent.node.directives.iter().any(|directive| directive.node.name.node == MODEL_DIRECTIVE);
            then {
                ctx.report_error(
                    vec![field.pos],
                    format!("Field {name} is reserved, you can't use it.", name = name),
                );

            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::CheckModelizedFieldReserved;
    use crate::rules::visitor::{visit, VisitorContext};
    use dynaql_parser::parse_schema;
    use serde_json as _;

    #[test]
    fn should_error_on_reserved_keyword() {
        let schema = r#"
            type Product @model {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Int!
                created_at: DateTime!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckModelizedFieldReserved, &mut ctx, &schema);

        assert!(!ctx.errors.is_empty(), "shouldn't be empty");
        assert_eq!(ctx.errors.len(), 1, "should have one error");
        assert_eq!(
            ctx.errors.get(0).unwrap().message,
            "Field created_at is reserved, you can't use it.",
            "should match"
        );
    }

    #[test]
    fn should_allow_reserved_keyword_on_non_model() {
        let schema = r#"
            type Product {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Int!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckModelizedFieldReserved, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should not have any error");
    }
}
