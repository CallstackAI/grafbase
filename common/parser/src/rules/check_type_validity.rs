//! ### What it does
//!
//! Check the types for fields are valid.
//! Fields valid are: Primitives, Basic Type.
//!
//! ### Why?
//!
//! To avoid having an invalid schema
use super::model_directive::MODEL_DIRECTIVE;
use super::visitor::{Visitor, VisitorContext};
use crate::utils::is_type_primitive;
use crate::utils::to_base_type_str;

pub struct CheckTypeValidity;

impl<'a> Visitor<'a> for CheckTypeValidity {
    fn enter_field(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        field: &'a async_graphql::Positioned<async_graphql_parser::types::FieldDefinition>,
        _parent_type: &'a async_graphql::Positioned<async_graphql_parser::types::TypeDefinition>,
    ) {
        let base_type = to_base_type_str(&field.node.ty.node.base);
        if is_type_primitive(&field.node) {
            return;
        }

        match ctx.types.get(&base_type) {
            Some(ty) => {
                if ty.node.directives.iter().any(|d| d.node.name.node == MODEL_DIRECTIVE) {
                    ctx.report_error(
                        vec![field.pos],
                        format!(
                            "Field `{name}` got an undefined type: `{ty}`.",
                            name = field.node.name.node,
                            ty = base_type
                        ),
                    );
                }
            }
            None => {
                ctx.report_error(
                    vec![field.pos],
                    format!(
                        "Field `{name}` got an undefined type: `{ty}`.",
                        name = field.node.name.node,
                        ty = base_type
                    ),
                );
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::rules::check_type_validity::CheckTypeValidity;
    use crate::rules::visitor::{visit, VisitorContext};
    use async_graphql_parser::parse_schema;
    use serde_json as _;

    #[test]
    fn should_not_error_with_basic_type() {
        let schema = r#"
            type Product @model {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Int!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckTypeValidity, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
    }

    #[test]
    fn should_not_error_with_custom_type() {
        let schema = r#"
            type Truc {
                name: String!
            }

            type Product {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Truc!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckTypeValidity, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should be empty");
    }

    #[test]
    fn should_error_with_model_type() {
        let schema = r#"
            type Truc @model {
                name: String!
            }

            type Product {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Truc!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckTypeValidity, &mut ctx, &schema);

        assert!(!ctx.errors.is_empty(), "shouldn't be empty");
        assert_eq!(ctx.errors.len(), 1, "should have one error");
        assert_eq!(
            ctx.errors.get(0).unwrap().message,
            "Field `__price` got a modelized type: `Truc` but modelized types are not valid as field yet.",
            "should match"
        );
    }

    #[test]
    fn should_error_with_invalid_types() {
        let schema = r#"
            type Product @model {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Url!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckTypeValidity, &mut ctx, &schema);

        assert!(!ctx.errors.is_empty(), "shouldn't be empty");
        assert_eq!(ctx.errors.len(), 1, "should have one error");
        assert_eq!(
            ctx.errors.get(0).unwrap().message,
            "Field `__price` got an undefined type: `Url`.",
            "should match"
        );
    }
}
