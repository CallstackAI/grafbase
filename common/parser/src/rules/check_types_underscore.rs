//! ### What it does
//!
//! Check that types input by the user don't begin with double underscores
//! on @models only.
//!
//! ### Why?
//!
//! We keep those types as internal types.

use super::visitor::{Visitor, VisitorContext};
use if_chain::if_chain;

pub struct CheckBeginsWithDoubleUnderscore;

impl<'a> Visitor<'a> for CheckBeginsWithDoubleUnderscore {
    fn enter_field(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        field: &'a dynaql::Positioned<dynaql_parser::types::FieldDefinition>,
        parent: &'a dynaql::Positioned<dynaql_parser::types::TypeDefinition>,
    ) {
        if_chain! {
            let name = &field.node.name.node;
            if name.starts_with("__");
            if parent.node.directives.iter().any(|directive| directive.is_model());
            then {
                ctx.report_error(
                    vec![field.pos],
                    format!("Field {name} shouldn't start with double underscore."),
                );

            }
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::rules::check_types_underscore::CheckBeginsWithDoubleUnderscore;
    use crate::rules::visitor::{visit, VisitorContext};
    use dynaql_parser::parse_schema;
    use serde_json as _;

    #[test]
    fn should_error_on_underscore() {
        let schema = r#"
            type Product @model {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Int!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckBeginsWithDoubleUnderscore, &mut ctx, &schema);

        assert!(!ctx.errors.is_empty(), "shouldn't be empty");
        assert_eq!(ctx.errors.len(), 1, "should have one error");
        assert_eq!(
            ctx.errors.get(0).unwrap().message,
            "Field __price shouldn't start with double underscore.",
            "should match"
        );
    }

    #[test]
    fn should_allow_underscore_on_non_model() {
        let schema = r#"
            type Product {
                id: ID!
                _name: String!
                """
                The product's price in $
                """
                __price: Int!
            }
            "#;

        let schema = parse_schema(schema).expect("");

        let mut ctx = VisitorContext::new(&schema);
        visit(&mut CheckBeginsWithDoubleUnderscore, &mut ctx, &schema);

        assert!(ctx.errors.is_empty(), "should not have any error");
    }
}
