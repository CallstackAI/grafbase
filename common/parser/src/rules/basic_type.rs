//! For basic types
//!
//! When a basic type is stubble uppon on the definition of the schema, if it
//! got no specialized behavior, we apply this behavior uppon it.
//!
use super::visitor::{Visitor, VisitorContext};
use crate::registry::add_input_type_non_primitive;
use crate::rules::cache_directive::CacheDirective;
use dynaql::registry::resolvers::transformer::Transformer;
use dynaql::registry::{self, MetaField};
use dynaql_parser::types::TypeKind;
use if_chain::if_chain;

pub struct BasicType;

impl<'a> Visitor<'a> for BasicType {
    fn enter_type_definition(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        type_definition: &'a dynaql::Positioned<dynaql_parser::types::TypeDefinition>,
    ) {
        let directives = &type_definition.node.directives;
        if_chain! {
            if !["Query", "Mutation"].contains(&type_definition.node.name.node.as_str());
            if !directives.iter().any(|directive| directive.is_model());
            if let TypeKind::Object(object) = &type_definition.node.kind;
            then {
                let type_name = type_definition.node.name.node.to_string();
                // If it's a modeled Type, we create the associated type into the registry.
                // Without more data, we infer it's from our modelization.
                ctx.registry.get_mut().create_type(|_| registry::ObjectType::new(
                    type_name.clone(),
                    object.fields.iter().map(|field| {
                        let name = field.name().to_string();
                        let mapped_name = field.mapped_name().map(ToString::to_string);

                        MetaField {
                            name: name.clone(),
                            mapped_name,
                            description: field.node.description.clone().map(|x| x.node),
                            ty: field.node.ty.clone().node.to_string().into(),
                            cache_control: CacheDirective::parse(&field.node.directives),
                            resolver: Transformer::select(&name).into(),
                            ..Default::default()
                        }
                    })
                )
                    .with_description(type_definition.node.description.clone().map(|x| x.node))
                    .with_cache_control(CacheDirective::parse(&type_definition.node.directives))
                    .into()
                ,&type_name,
                &type_name
                );

                // If the type is a non primitive and also not modelized, it means we need to
                // create the Input version of it.
                // If the input is non used by other queries/mutation, it'll be removed from the
                // final schema.
                add_input_type_non_primitive(ctx, object, &type_name);
            }
        }
    }
}
