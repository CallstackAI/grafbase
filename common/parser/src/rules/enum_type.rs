//! For enum types
//!
//! There is no specialied behavior for enum right now.
//!
//! TODO: Manage deprecation
use super::visitor::{Visitor, VisitorContext};
use async_graphql::indexmap::IndexMap;
use async_graphql::registry::MetaEnumValue;
use async_graphql::registry::MetaType;
use async_graphql_parser::types::TypeKind;
use if_chain::if_chain;

pub struct EnumType;

impl<'a> Visitor<'a> for EnumType {
    fn enter_type_definition(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        type_definition: &'a async_graphql::Positioned<async_graphql_parser::types::TypeDefinition>,
    ) {
        if_chain! {
            if let TypeKind::Enum(enum_ty) = &type_definition.node.kind;
            then {
                let type_name = type_definition.node.name.node.to_string();
                ctx.registry.get_mut().create_type(&mut |_| MetaType::Enum {
                    name: type_name.clone(),
                    description: type_definition.node.description.clone().map(|x| x.node),
                    visible: None,
                    rust_typename: type_name.clone(),
                    enum_values: {
                        let mut values = IndexMap::new();
                        for v in &enum_ty.values {
                            let enum_value = &v.node.value.node;
                            values.insert(
                                enum_value.to_string(),
                                MetaEnumValue {
                                    name: enum_value.to_string(),
                                    description: v.node.description.clone().map(|x| x.node),
                                    deprecation: async_graphql::registry::Deprecation::NoDeprecated,
                                    visible: None,
                                }
                                );
                        }
                        values
                    }
                }, &type_name, &type_name);
            }
        }
    }
}
