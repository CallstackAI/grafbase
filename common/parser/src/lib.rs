use async_graphql::registry::Registry;
use async_graphql_parser::{parse_schema, Error as ParserError};
use quick_error::quick_error;
use rules::basic_type::BasicType;
use rules::check_field_not_reserved::CheckModelizedFieldReserved;
use rules::check_type_validity::CheckTypeValidity;
use rules::check_types_underscore::CheckBeginsWithDoubleUnderscore;
use rules::enum_type::EnumType;
use rules::model_directive::ModelDirective;
use rules::relations::relations_rules;
use rules::visitor::{visit, RuleError, Visitor, VisitorContext};

mod registry;
mod rules;
mod utils;

quick_error! {
    #[derive(Debug)]
    pub enum Error {
        Parser(err: ParserError) {
            from()
            source(err)
            display("{}", err)
        }
        Validation(err: Vec<RuleError>) {
            from()
            display("{:?}", err)
        }
    }
}

/// Transform the input schema into a Registry
pub fn to_registry<S: AsRef<str>>(input: S) -> Result<Registry, Error> {
    let mut rules = rules::visitor::VisitorNil
        .with(CheckBeginsWithDoubleUnderscore)
        .with(CheckModelizedFieldReserved)
        .with(CheckTypeValidity)
        .with(ModelDirective)
        .with(BasicType)
        .with(EnumType)
        .with(relations_rules());

    let schema = parse_schema(format!("{}\n{}", rules.directives(), input.as_ref()))?;

    let mut ctx = VisitorContext::new(&schema);
    visit(&mut rules, &mut ctx, &schema);

    if !ctx.errors.is_empty() {
        return Err(ctx.errors.into());
    }

    let reg = ctx.finish();
    Ok(reg)
}

#[cfg(test)]
mod tests {
    use async_graphql::Schema;
    use serde_json as _;

    #[test]
    fn test_simple_product() {
        let result = super::to_registry(
            r#"
            type Product @model {
                id: ID!
                name: String!
                """
                The product's price in $
                """
                price: Int!
            }
            "#,
        )
        .unwrap();

        let reg_string = serde_json::to_value(&result).unwrap();
        let sdl = Schema::new(result).sdl();

        insta::assert_json_snapshot!(reg_string);
        insta::assert_snapshot!(sdl);
    }

    #[test]
    fn test_simple_todo() {
        let result = super::to_registry(
            r#"
            type Todo @model {
              id: ID!
              content: String!
              author: Author
            }

            type Author {
              name: String!
              lastname: String!
              pseudo: String
              truc: Truc!
            }

            type Truc {
              name: String!
            }
            "#,
        )
        .unwrap();

        let reg_string = serde_json::to_value(&result).unwrap();
        let sdl = Schema::new(result).sdl();

        insta::assert_json_snapshot!(reg_string);
        insta::assert_snapshot!(sdl);
    }

    #[test]
    fn test_simple_todo_from_template() {
        let result = super::to_registry(
            r#"
            type TodoList @model {
              id: ID!
              title: String!
              todos: [Todo]
            }

            type Todo @model {
              id: ID!
              title: String!
              complete: Boolean
            }
            "#,
        )
        .unwrap();

        let sdl = Schema::new(result).sdl();

        insta::assert_snapshot!(sdl);
    }

    #[test]
    fn test_simple_todo_with_vec() {
        let result = super::to_registry(
            r#"
            type Todo @model {
              id: ID!
              content: String!
              authors: [Author]
            }

            type Author {
              name: String!
              lastname: String!
              pseudo: String
              truc: Truc!
            }

            type Truc {
              name: String!
            }
            "#,
        )
        .unwrap();

        let reg_string = serde_json::to_value(&result).unwrap();
        let sdl = Schema::new(result).sdl();

        insta::assert_json_snapshot!(reg_string);
        insta::assert_snapshot!(sdl);
    }

    #[test]
    fn test_simple_todo_with_enum() {
        let result = super::to_registry(
            r#"
            """
            A TodoType
            """
            enum TodoType {
              TODO1

              """
              A Type 2 for TODO
              """
              TODO2
            }

            type Todo @model {
              id: ID!
              content: String!
              authors: [Author]
              ty: TodoType!
            }

            type Author {
              name: String!
              lastname: String!
              pseudo: String
              truc: Truc!
            }

            type Truc {
              name: String!
            }
            "#,
        )
        .unwrap();

        let reg_string = serde_json::to_value(&result).unwrap();
        let sdl = Schema::new(result).sdl();

        insta::assert_json_snapshot!(reg_string);
        insta::assert_snapshot!(sdl);
    }

    #[test]
    fn test_simple_post_with_relation() {
        let result = super::to_registry(
            r#"
            enum Country {
              FRANCE
              NOT_FRANCE
            }

            type Blog @model {
              id: ID!
              posts: [Post] 
              owner: Author!
            }

            type Post @model {
              id: ID!
              content: String!
              authors: [Author] @relation(name: "published")
            }

            type Author @model {
              id: ID!
              name: String!
              lastname: String!
              country: Country!
              posts: [Post] @relation(name: "published")
            }
            "#,
        )
        .unwrap();

        let reg_string = serde_json::to_value(&result).unwrap();
        let sdl = Schema::new(result).sdl();

        insta::assert_json_snapshot!(reg_string);
        insta::assert_snapshot!(sdl);
        assert_eq!(false, true);
    }
}
