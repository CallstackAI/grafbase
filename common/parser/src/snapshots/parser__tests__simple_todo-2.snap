---
source: common/parser/src/lib.rs
assertion_line: 112
expression: sdl
---
type Author {
	name: String!
	lastname: String!
	pseudo: String
	truc: Truc!
}
"""
Author input type.
"""
input AuthorInput {
	name: String!
	lastname: String!
	pseudo: String
	truc: TrucInput!
}
type Mutation {
	"""
	Create a Todo
	"""
	todoCreate(input: TodoCreationInput!): TodoCreatePayload
	"""
	Delete a Todo by ID
	"""
	todoDelete(id: ID!): TodoDeletePayload
}
type PageInfo {
	hasPreviousPage: Boolean!
	hasNextPage: Boolean!
	startCursor: String
	endCursor: String
}
type Query {
	"""
	Get a Todo by ID
	"""
	todo(id: ID!): Todo
	"""
	Paginated query to fetch the whole list of `Todo`.
	"""
	todoCollection(after: String, before: String, first: Int, last: Int): TodoConnection
}
type Todo {
	id: ID!
	content: String!
	author: Author
}
type TodoConnection {
	"""
	Information to aid in pagination
	"""
	pageInfo: PageInfo!
	edges: [TodoEdge]
}
type TodoCreatePayload {
	todo: Todo
}
"""
Input to create a new Todo
"""
input TodoCreationInput {
	content: String!
	author: AuthorInput
}
type TodoDeletePayload {
	deletedId: ID!
}
type TodoEdge {
	node: Todo!
	cursor: String!
}
type Truc {
	name: String!
}
"""
Truc input type.
"""
input TrucInput {
	name: String!
}
schema {
	query: Query
	mutation: Mutation
}

